lib/index.js

```diff
@@ -479,6 +488,83 @@ Socket.prototype.send = function(msg, flags) {
   return this;
 };
 
+/**
+ * Read one message from socket.
+ *
+ * @return {Array} multi-part message or null if socket doesn't have any
+ * @api public
+ */
+Socket.prototype.read = function() {
+  var message = [], flags;
+
+  if (this._zmq.state !== zmq.STATE_READY) {
+    return null;
+  }
+
+  flags = this._zmq.getsockopt(zmq.ZMQ_EVENTS);
+
+  if (flags & zmq.ZMQ_POLLIN) {
+    do {
+      message.push(this._zmq.recv());
+    } while (this._zmq.getsockopt(zmq.ZMQ_RCVMORE));
+
+    return message;
+  }
+
+  return null;
+}
+
+/**
+ * Send the given `msg`.
+ *
+ * @param {String|Buffer|Array} msg
+ * @param {Function} callback function (optional)
+ * @return {Socket} for chaining
+ * @api public
+ */
+Socket.prototype.write = function (parts, callback) {
+  var self = this,
+    i = 0,
+    len = parts.length,
+    flags = this._zmq.getsockopt(zmq.ZMQ_EVENTS),
+    part;
+
+  if (!Array.isArray(parts)) {
+    parts = [parts];
+  }
+
+  // socket is not ready to accept messages
+  if (!(flags & zmq.ZMQ_POLLOUT)) {
+    server.once('ready', function () {
+      self.write(parts, callback);
+    });
```

Comment:
Callback is needed for the writable stream. So you will either pile up closures or you pile up data in arrays (streams in object mode would do instead of _outgoing). 

Bottom line there is no way that we can take responsibility from developer if there is a mem leak due to high injection rate. 

PS.> In any other language this would block according to 0mq spec.